state/store.go

```diff
@@ -481,9 +489,11 @@ func (store dbStore) PruneABCIResponses(targetRetainHeight int64, forceCompact b
 		return pruned + batchPruned, targetRetainHeight, err
 	}
 
-	if forceCompact && store.Compact {
-		if pruned+batchPruned >= store.CompactionInterval || targetRetainHeight-lastRetainHeight >= store.CompactionInterval {
-			err = store.db.Compact(nil, nil)
+	store.StoreOptions.ResultsToCompact += uint64(pruned + batchPruned)
+	if forceCompact && store.Compact && store.StoreOptions.ResultsToCompact >= (uint64)(store.StoreOptions.CompactionInterval) {
```

Comment:
I would structure this differently. If `store.Compact` is false, we don't want `store.StoreOptions.ResultsToCompact` to be growing forever and eventually wrapping.
I would add an outer IF with checking `store.Compact` only. And, if true, then we update `store.StoreOptions.ResultsToCompact` and do the inner IF with the rest of the conditions.